amgcl/preconditioner/cpr.hpp

```diff
@@ -150,11 +150,22 @@ class cpr {
         }
 
         template <class Matrix>
-        void update_sprecond(
+        void partial_update(
                 const Matrix &K,
-                const backend_params &bprm = backend_params())
+                bool update_transfer_ops = true,
+                const backend_params &bprm = backend_params()
+              )
         {
+            // Update global preconditioner
             S = std::make_shared<SPrecond>(K, prm.sprecond, bprm);
+            if(update_transfer_ops){
+              // Update transfer operator Fpp
+              update_transfer(
+                  std::make_shared<build_matrix>(K),
+                  bprm,
+                  std::integral_constant<bool, math::static_rows<value_type>::value == 1>()
+                );
+            }
```

Comment:
This will save some work on converting from generic matrix to a `shared_ptr<build_matrix>` (otherwise SPrecond will do the same conversion in its constructor):
```cpp
            auto K_ptr = std::make_shared<build_matrix>(K);

            // Update global preconditioner
            S = std::make_shared<SPrecond>(K_ptr, prm.sprecond, bprm);

            if(update_transfer_ops){
              // Update transfer operator Fpp
              update_transfer(K_ptr, bprm,
                  std::integral_constant<bool, math::static_rows<value_type>::value == 1>()
                );
            }
```

Same applies for `cpr_drs`.

Reply:
I think this is fixed in the latest commits!